pkg/agent/billing.go

```diff
@@ -0,0 +1,182 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/google/uuid"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	vmapi "github.com/neondatabase/neonvm/apis/neonvm/v1"
+
+	"github.com/neondatabase/autoscaling/pkg/billing"
+	"github.com/neondatabase/autoscaling/pkg/util"
+)
+
+type BillingConfig struct {
+	URL                 string `json:"url"`
+	CollectEverySeconds uint   `json:"collectEverySeconds"`
+	PushEverySeconds    uint   `json:"pushEverySeconds"`
+	PushTimeoutSeconds  uint   `json:"pushTimeoutSeconds"`
+}
+
+type billingMetricsState struct {
+	historical map[billingMetricsKey]vmMetricsHistory
+	present    map[billingMetricsKey]vmMetrics
+	lastTime   *time.Time
+}
+
+type billingMetricsKey struct {
+	uid        types.UID
+	endpointID string
+}
+
+type vmMetricsHistory struct {
+	slices []metricsTimeSlice
+}
+
+type metricsTimeSlice struct {
+	metrics   vmMetrics
+	startTime time.Time
+	endTime   time.Time
+}
+
+type vmMetrics struct {
+	cpu uint16
```

Comment:
From discussion with @tychoish — this shouldn't be "cpu at a given time", instead should be "cpu seconds consumed over a timeslice". Will fix.

Reply:
fixed in latest commit